src/gotcha_dl.c

```diff
@@ -5,7 +5,137 @@
+                    addrs->found = 1;
+                    addrs->lmap = current;
+                    break;
+                }
+                current = current->l_next;
+            }
+        }
+    }
+    return 0;
+}
+/**
+ * This function looks for all symbols defined in rela or rel.
+ *
+ * @param name, symbol name
+ * @param lmap, library where we search for symbol.
+ * @return symbol pointer
+ */
+int dl_lookup_symbol_x(const char *name, struct link_map *lmap, void** symbol) {
+    long result;
+    INIT_DYNAMIC(lmap)
+    if (gnu_hash) {
+        debug_printf(3, "Checking GNU hash for %s in %s\n", name, LIB_NAME(lmap));
+        struct gnu_hash_header * header;
+        if (is_rela) header = (struct gnu_hash_header *) (gnu_hash);
+        else header = (struct gnu_hash_header *) (lmap->l_addr + gnu_hash);
+        result = lookup_gnu_hash_symbol(name, symtab, strtab, header);
+        if (result != -1) {
+            if (GOTCHA_CHECK_VISIBILITY(symtab[result])) {
+                *symbol = (void *) (symtab[result].st_value + lmap->l_addr);
+                return 0;
+            } else {
+                debug_printf(3, "Symbol %s found but not exported in %s\n", name, LIB_NAME(lmap));
+            }
+        } else {
+            debug_printf(3, "%s not found in %s\n", name, LIB_NAME(lmap));
+        }
+    } else {
+        debug_printf(3, "Library %s does not export symbols\n", LIB_NAME(lmap));
+    }
+    return -1;
+}
+
+/**
+ * Implement the logic of _dl_sym
+ * 1. find the caller library using the program headers.
+ * 2. find the first library which has the symbol for RTLD_DEFAULT
+ * 3. find the second library which has the symbol for RTLD_NEXT
+ * @param handle, handle for dl operation
+ * @param name, name of the symbol
+ * @param who, the virtual address of the caller
+ * @return symbol pointer
+ */
+
+static void * gotchas_dlsym_rtld_next_lookup(void *handle, const char *name, void *who) {
+    ElfW(Addr) caller = (ElfW(Addr)) who;
+    /* Iterative over the library headers and find the caller
+     * the address of the caller is set in addrs->library_laddr
+     **/
+    struct Addrs addrs;
+    addrs.found = 0;
+    addrs.lookup_addr = caller;
+    addrs.lmap = _r_debug.r_map;
+    void* symbol;
+    dl_iterate_phdr(lib_header_callback, &addrs);
+    if (addrs.lmap->l_next) {
+        struct link_map *lib_iter;
+        int found_caller = 1;
+        int found_sym = 0;
+        for (lib_iter = addrs.lmap->l_next; lib_iter != 0; lib_iter = lib_iter->l_next) {
+            /* lookup symbol on the next lib or next-to-next lib which has symbol
+             * This selection depends on RTLD_DEFAULT or RTLD_NEXT
+             **/
+            int res = dl_lookup_symbol_x(name, lib_iter, &symbol);
+            if (res != -1) {
+                if (handle == RTLD_DEFAULT) {
+                    return symbol;
+                } else if (found_sym && handle == RTLD_NEXT){
+                    return symbol;
```

Comment:
Rather than return the symbol, we should return the link_map that contains the symbol. Also see the below comment on what to do with that link_map.

Reply:
returning link map now.